app/javascript/components/laboratory/Laboratory.js

```diff
@@ -16,23 +17,41 @@ class Laboratory extends React.Component {
       specimen_collection: this.props.lab.specimen_collection,
       report: this.props.lab.report,
       result: this.props.lab.result || '',
+      reportInvalid: false,
     };
     this.toggleModal = this.toggleModal.bind(this);
     this.handleChange = this.handleChange.bind(this);
     this.submit = this.submit.bind(this);
   }
 
   toggleModal() {
-    let current = this.state.showModal;
-    this.setState({
-      showModal: !current,
+    this.setState(state => {
+      return {
+        showModal: !state.showModal,
+        loading: false,
+        lab_type: this.props.lab.lab_type || '',
+        specimen_collection: this.props.lab.specimen_collection,
+        report: this.props.lab.report,
+        result: this.props.lab.result || '',
+        reportInvalid: false,
+      };
     });
   }
 
   handleChange(event) {
     this.setState({ [event.target.id]: event.target.value });
   }
 
+  handleDateChange(field, date) {
+    this.setState({ [field]: date }, () => {
```

Comment:
Why not just make this one `setState`?

Reply:
because the values `state.report` and `state.specimen_collection` used in the inner block depend on those fields being updated first